sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -21,38 +21,232 @@
+ * for 1 hour.
+ *
+ * <pre>{@code
+ * PCollection<Metadata> matches = p.apply(FileIO.match()
+ *     .filepattern("...")
+ *     .continuously(
+ *       Duration.standardSeconds(30), afterTimeSinceNewOutput(Duration.standardHours(1))));
+ * }</pre>
+ *
+ * <h3>Example: Matching a PCollection of filepatterns arriving from Kafka</h3>
+ *
+ * <p>This example reads filepatterns from Kafka and matches each one as it arrives, producing again
+ * an unbounded {@code PCollection<Metadata>}, and failing in case the filepattern doesn't match
+ * anything.
+ *
+ * <pre>{@code
+ * PCollection<String> filepatterns = p.apply(KafkaIO.read()...);
+ *
+ * PCollection<Metadata> matches = filepatterns.apply(FileIO.matchAll()
+ *     .withEmptyMatchTreatment(DISALLOW));
+ * }</pre>
+ *
+ * <h2>Reading files</h2>
+ *
+ * <p>{@link #readMatches} converts each result of {@link #match} or {@link #matchAll} to a {@link
+ * ReadableFile} that is convenient for reading a file's contents, optionally decompressing it.
+ *
+ * <h3>Example: Returning filenames and contents of compressed files matching a filepattern</h3>
+ *
+ * <p>This example matches a single filepattern and returns {@code KVs} of filenames and their
+ * contents as {@code String}, decompressing each file with GZIP.
+ *
+ * <pre>{@code
+ * PCollection<KV<String, String>> filesAndContents = p
+ *     .apply(FileIO.match().filepattern("hdfs://path/to/*.gz"))
+ *     // withCompression can be omitted - by default compression is detected from the filename.
+ *     .apply(FileIO.readMatches().withCompression(GZIP))
+ *     .apply(MapElements
+ *         // uses imports from TypeDescriptors
+ *         .into(KVs(strings(), strings()))
+ *         .via((ReadableFile f) -> KV.of(
+ *             f.getMetadata().resourceId().toString(), f.readFullyAsUTF8String())));
+ * }</pre>
+ *
+ * <h2>Writing files</h2>
+ *
+ * <p>{@link #write} and {@link #writeDynamic} write elements from a {@link PCollection} of a given
+ * type to files, using a given {@link Sink} to write a set of elements to each file. The collection
+ * can be bounded or unbounded - in either case, writing happens by default per window, and the
+ * amount of data in each window is finite, so a finite number of files ("shards") are written for
+ * each window. There are several aspects to this process:
+ *
+ * <ul>
+ *   <li><b>How many shards are generated per window:</b> This is controlled by <i>sharding</i>,
+ *       using {@link Write#withNumShards} or {@link Write#withSharding}. By default, every bundle
+ *       in the input {@link PCollection} will produce a shard, so the number of shards will vary
+ *       based on runner behavior, though at least 1 shard will always be produced for every
+ *       non-empty window. Note that setting a fixed number of shards can hurt performance: it adds
+ *       an additional {@link GroupByKey} to the pipeline.
+ *   <li><b>How the shards are named:</b> This is controlled by a {@link Write.FilenamePolicy}:
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
More how the files are named, as a single shard might consist of amny files.